RenderMathMLRoot does not reset radical operator when children change dynamically

Ahmad-S792 opened the pull request:
#### 498cdeb54915897996f55c8eb5cac655021a74ce
<pre>
RenderMathMLRoot does not reset radical operator when children change dynamically
<a href="https://bugs.webkit.org/show_bug.cgi?id=304194">https://bugs.webkit.org/show_bug.cgi?id=304194</a>
<a href="https://rdar.apple.com/166556627">rdar://166556627</a>

Reviewed by Frédéric Wang.

When children are dynamically removed from msqrt or mroot elements, the radical
operator retains its previously stretched width from the larger content,
causing the element to be wider than necessary for the remaining children.

The issue occurs because the renderer is not notified when the child list
changes. While styleDidChange() properly resets m_radicalOperator for style
changes, there was no corresponding mechanism to reset it when children are
added or removed from the DOM.

Fix by overriding childrenChanged() in MathMLRootElement to notify the
renderer when the child list changes. The renderer resets m_radicalOperator
and marks itself as needing layout and preferred width recalculation, ensuring
width calculations reflect the current children rather than stale cached state.

* Source/WebCore/mathml/MathMLRootElement.cpp:
(WebCore::MathMLRootElement::childrenChanged):
* Source/WebCore/mathml/MathMLRootElement.h:
* Source/WebCore/rendering/mathml/RenderMathMLRoot.cpp:
(WebCore::RenderMathMLRoot::styleDidChange):
* Source/WebCore/rendering/mathml/RenderMathMLRoot.h:

&gt; Progressions:
* LayoutTests/imported/w3c/web-platform-tests/mathml/relations/html5-tree/dynamic-childlist-002-expected.txt:
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/mathml/relations/html5-tree/dynamic-childlist-001-expected.txt:
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/mathml/relations/html5-tree/dynamic-childlist-002-expected.txt:
* LayoutTests/platform/ios/imported/w3c/web-platform-tests/mathml/relations/html5-tree/dynamic-childlist-001-expected.txt:
* LayoutTests/platform/ios/imported/w3c/web-platform-tests/mathml/relations/html5-tree/dynamic-childlist-002-expected.txt:
* LayoutTests/platform/mac/imported/w3c/web-platform-tests/mathml/relations/html5-tree/dynamic-childlist-001-expected.txt:

Canonical link: <a href="https://commits.webkit.org/304822@main">https://commits.webkit.org/304822@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/5054a5a04ee2033685f031e6426158255d5f9a64

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/136576 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/8935 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/47863 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/144300 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/89549 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/3a75bf2f-ba37-46bf-af0d-a38c4af14c0e) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/138448 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/9631 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/8781 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/104433 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/75005 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/b9a70548-097e-427a-968d-464c943350bb) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/139521 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/7029 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/122370 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/85268 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/fdb0eec8-7a19-45a9-9665-11e44647ddf9) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/6673 "Passed tests") | [✅ 🧪 api-mac-debug](https://ews-build.webkit.org/#/builders/165/builds/4346 "Passed tests") | [✅ 🛠 wpe-cairo-libwebrtc](https://ews-build.webkit.org/#/builders/166/builds/4896 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/115984 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/40562 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/147058 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/8619 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/41137 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/112777 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/8637 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/7245 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/113116 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/28737 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/6600 "Passed tests") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/118667 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/62652 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/8667 "Built successfully") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/36721 "Passed tests") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/8386 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/72233 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/8607 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/8459 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

fred-wang commented on Source/WebCore/rendering/mathml/RenderMathMLRoot.cpp:
```diff
@@ -166,6 +166,9 @@ void RenderMathMLRoot::computePreferredLogicalWidths()
 
+    // Reset the radical operator to ensure it's sized for current content.
+    m_radicalOperator.reset(style());
+
```
Can you explain why the preferred width depends on the list of children? IIUC, MathOperator will always use the max width of available glyphs / assembly.

`RenderMathMLRoot::styleDidChange()` already handles the style change. For child list changes, we probably want to override `MathMLRowElement::childrenChanged` in `MathMLRootElement` so that it tells the render class to reset `m_radicalOperator`. And maybe set `RenderObject::setNeedsLayout()` (or `RenderObject::setNeedsLayoutAndPreferredWidthsUpdate()`) ?

fred-wang replied:
The WPT tests fixed, seem related to the layout, not to the min-max width calculation.

Ahmad-S792 replied:
@fred-wang - You were right, I addressed this, appreciate if you can review now. Thanks!

fred-wang commented on Source/WebCore/mathml/MathMLRootElement.cpp:
```diff
@@ -63,6 +63,18 @@ RenderPtr<RenderElement> MathMLRootElement::createElementRenderer(RenderStyle&&
+    MathMLRowElement::childrenChanged(change);
+
+    if (CheckedPtr renderer = this->renderer()) {
+        if (CheckedPtr rootRenderer = dynamicDowncast<RenderMathMLRoot>(*renderer)) {
```
does a single `if (CheckedPtr rootRenderer = dynamicDowncast<RenderMathMLRoot>(this->renderer()))` work?

Ahmad-S792 replied:
Done. It compiles locally, will wait for EWS before merging.

fred-wang reviewed (approved):
Thanks. LGTM

Ahmad-S792 committed:
RenderMathMLRoot does not reset radical operator when children change dynamically
https://bugs.webkit.org/show_bug.cgi?id=304194
rdar://166556627

Reviewed by Frédéric Wang.

When children are dynamically removed from msqrt or mroot elements, the radical
operator retains its previously stretched width from the larger content,
causing the element to be wider than necessary for the remaining children.

The issue occurs because the renderer is not notified when the child list
changes. While styleDidChange() properly resets m_radicalOperator for style
changes, there was no corresponding mechanism to reset it when children are
added or removed from the DOM.

Fix by overriding childrenChanged() in MathMLRootElement to notify the
renderer when the child list changes. The renderer resets m_radicalOperator
and marks itself as needing layout and preferred width recalculation, ensuring
width calculations reflect the current children rather than stale cached state.

* Source/WebCore/mathml/MathMLRootElement.cpp:
(WebCore::MathMLRootElement::childrenChanged):
* Source/WebCore/mathml/MathMLRootElement.h:
* Source/WebCore/rendering/mathml/RenderMathMLRoot.cpp:
(WebCore::RenderMathMLRoot::styleDidChange):
* Source/WebCore/rendering/mathml/RenderMathMLRoot.h:

> Progressions:
* LayoutTests/imported/w3c/web-platform-tests/mathml/relations/html5-tree/dynamic-childlist-002-expected.txt:
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/mathml/relations/html5-tree/dynamic-childlist-001-expected.txt:
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/mathml/relations/html5-tree/dynamic-childlist-002-expected.txt:
* LayoutTests/platform/ios/imported/w3c/web-platform-tests/mathml/relations/html5-tree/dynamic-childlist-001-expected.txt:
* LayoutTests/platform/ios/imported/w3c/web-platform-tests/mathml/relations/html5-tree/dynamic-childlist-002-expected.txt:
* LayoutTests/platform/mac/imported/w3c/web-platform-tests/mathml/relations/html5-tree/dynamic-childlist-001-expected.txt:

Canonical link: https://commits.webkit.org/304822@main
